File: java/src/main/java/com/google/cloud/crypto/tink/Aead.java
// Copyright 2017 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

package com.google.cloud.crypto.tink;

import com.google.common.util.concurrent.ListenableFuture;
import java.security.GeneralSecurityException;

/**
 * The interface for authenticated encryption with additional authenticated data.
 * Implementations of this interface are secure against adaptive chosen ciphertext attacks.
 * Encryption with additional data ensures authenticity and integrity of that data,
 * but not its secrecy. (see RFC 5116, https://tools.ietf.org/html/rfc5116)
 */
public interface Aead {
  /**
   * Encrypts {@code plaintext} with {@code additionalData} as additional authenticated data.
   * The resulting ciphertext allows for checking authenticity and integrity
   * of additional data ({@code additionalData}), but does not guarantee its secrecy.
   *
   * @return resulting ciphertext.
   */
  byte[] encrypt(final byte[] plaintext, final byte[] additionalData)
      throws GeneralSecurityException;

  /**
   * Decrypts {@code ciphertext} with {@code additionalData} as additional authenticated data.
   * The decryption verifies the authenticity and integrity of the additional data,
   * but there are no guarantees wrt. secrecy of that data.
   *
   * @return resulting plaintext.
   */
  byte[] decrypt(final byte[] ciphertext, final byte[] additionalData)
      throws GeneralSecurityException;

  /**
   * Encrypts {@code plaintext} with {@code additionalData} as additional authenticated data.
   * The resulting ciphertext allows for checking the authenticity and integrity
   * of the additional data, but does not guarantee its secrecy.
   *
   * If the encryption fails for some reason then future's {@code get()} method
   * will return {@code ExecutionException} whose {@code getCause()} method returns
   * {@code GeneralSecurityException}.
   *
   * @return a {@code ListenableFuture} that holds the resulting ciphertext.
   */
  ListenableFuture<byte[]> asyncEncrypt(final byte[] plaintext, final byte[] additionalData);

  /**
   * Decrypts {@code ciphertext} with {@code additionalData} as additional authenticated data.
   * The decryption verifies the authenticity and integrity of the additional data,
   * but there are no guarantees wrt. secrecy of that data.
   *
   * If the ciphertext is invalid or unauthenticated, or if the decryption fails for some
   * other reason, then future's {@code get()} method will return {@code ExecutionException}
   * whose {@code getCause()} method returns {@code GeneralSecurityException}.
   *
   * @return a {@code ListenableFuture} that holds the resulting plaintext.
   */
  ListenableFuture<byte[]> asyncDecrypt(final  byte[] ciphertext, final  byte[] additionalData);
}

File: python/tink/cc/pybind/aead.cc
// Copyright 2019 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
///////////////////////////////////////////////////////////////////////////////

#include "tink/cc/pybind/aead.h"

#include <string>
#include <utility>

#include "pybind11/pybind11.h"
#include "tink/aead.h"
#include "tink/util/statusor.h"
#include "tink/cc/pybind/tink_exception.h"

namespace crypto {
namespace tink {

using pybind11::google_tink::TinkException;

void PybindRegisterAead(pybind11::module* module) {
  namespace py = pybind11;
  py::module& m = *module;

  // TODO(b/146492561): Reduce the number of complicated lambdas.
  py::class_<Aead>(
      m, "Aead",
      "The interface for authenticated encryption with associated data. "
      "Implementations of this interface are secure against adaptive "
      "chosen ciphertext attacks.  Encryption with associated data ensures "
      "authenticity and integrity of that data, but not its secrecy. "
      "(see RFC 5116, https://tools.ietf.org/html/rfc5116)")

      .def(
          "encrypt",
          [](const Aead &self, const py::bytes &plaintext,
             const py::bytes &associated_data) -> py::bytes {
            util::StatusOr<std::string> result = self.Encrypt(
                std::string(plaintext), std::string(associated_data));
            if (!result.ok()) {
              throw TinkException(result.status());
            }
            return *std::move(result);
          },
          py::arg("plaintext"), py::arg("associated_data"),
          "Encrypts 'plaintext' with 'associated_data' as associated data, "
          "and returns the resulting ciphertext. "
          "The ciphertext allows for checking authenticity and integrity "
          "of the associated data, but does not guarantee its secrecy.")
      .def(
          "decrypt",
          [](const Aead &self, const py::bytes &ciphertext,
             const py::bytes &associated_data) -> py::bytes {
            // TODO(b/145925674)
            util::StatusOr<std::string> result = self.Decrypt(
                std::string(ciphertext), std::string(associated_data));
            if (!result.ok()) {
              throw TinkException(result.status());
            }
            return *std::move(result);
          },
          py::arg("ciphertext"), py::arg("associated_data"),
          "Decrypts 'ciphertext' with 'associated_data' as associated data, "
          "and returns the resulting plaintext. "
          "The decryption verifies the authenticity and integrity "
          "of the associated data, but there are no guarantees wrt. secrecy "
          "of that data.");
}

}  // namespace tink
}  // namespace crypto
